sql: issue GetRequests from SQL when possible

jordanlewis opened the pull request:
Closes #46758.

Previously, SQL unconditionally emitted ScanRequests for every read that
it produced, even if we were completely sure that the scan could contain
only a single key that was fully known at request time.

This was less efficient than necessary, because ScanRequests cause the
KV layer to do more work than GetRequests.

Here are the reasons why GetRequests are more efficient:

- MVCCGet, unlike MVCCScan, is able to use a prefix iterator, which
  allows it to make use of RocksDB/Pebble bloom filters. There may also
  be other wins at the storage layer, like avoiding prefetching
- various data structures have optimizations and fast-paths for
  single-key operations (e.g. timestamp cache, latch manager, refresh
  span tracking)
- various code paths are simpler in the point operation case
- a point operation implies only needing a single key allocation instead
  of a start and end key allocation
- a point operation implies only needing to return up to a single
  result, which means that we can avoid some indirection in various
  places
- A scan over a span that has at most 1 key, but does not know it, needs
  to iterate over all the versions (or eventually do a second expensive
  seek), looking for the non-existent next key in the span. A Get can do
  a single seek, read the key, and be done.

Release note (performance improvement): SQL will now emit GetRequests
when possible to KV instead of always emitting ScanRequests. This
manifests as a modest performance improvement for some workloads.

jordanlewis committed:
sql: issue GetRequests from SQL when possible

Previously, SQL unconditionally emitted ScanRequests for every read that
it produced, even if we were completely sure that the scan could contain
only a single key that was fully known at request time.

This was less efficient than necessary, because ScanRequests cause the
KV layer to do more work than GetRequests.

Here are the reasons why GetRequests are more efficient:

- MVCCGet, unlike MVCCScan, is able to use a prefix iterator, which
  allows it to make use of RocksDB/Pebble bloom filters. There may also
  be other wins at the storage layer, like avoiding prefetching
- various data structures have optimizations and fast-paths for
  single-key operations (e.g. timestamp cache, latch manager, refresh
  span tracking)
- various code paths are simpler in the point operation case
- a point operation implies only needing a single key allocation instead
  of a start and end key allocation
- a point operation implies only needing to return up to a single
  result, which means that we can avoid some indirection in various
  places
- A scan over a span that has at most 1 key, but does not know it, needs
  to iterate over all the versions (or eventually do a second expensive
  seek), looking for the non-existent next key in the span. A Get can do
  a single seek, read the key, and be done.

Release note (performance improvement): SQL will now emit GetRequests
when possible to KV instead of always emitting ScanRequests. This
manifests as a modest performance improvement for some workloads.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61583)
<!-- Reviewable:end -->

jordanlewis commented:
So far, this looks to make almost no difference, but I haven't spent time doing a careful benchmark, and I don't know the best/worst cases to test against.

I ran KV95 on a single gceworker node for 60 seconds on empty data directories before and after the patch.

After the patch:

```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   60.0s        0        2039148        33985.5      0.9      0.7      1.8      5.0     35.7  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   60.0s        0         106947         1782.4     10.2     10.0     17.8     21.0     41.9  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
   60.0s        0        2146095        35767.9      1.3      0.7      6.3     13.6     41.9
```


Before the patch:

```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   60.0s        0        2024099        33734.7      0.9      0.7      2.0      5.0     35.7  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   60.0s        0         106675         1777.9      9.9      9.4     16.8     21.0     39.8  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
   60.0s        0        2130774        35512.6      1.3      0.7      6.0     13.1     39.8
```

yuzefovich reviewed (commented):
Pretty cool stuff! It's the first time I looked into this code, but I still have some comments :)

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MV48MhoDsAVL-26jhfZ:b-km0vl5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @nvanbenschoten, and @solongordon)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MV466tf7YxKxzWpZR8q:-MV466tf7YxKxzWpZR8r:b-hewcyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a5ac6896e887fa5682cb5306b65c94de1b3373/pkg/kv/kvclient/kvcoord/dist_sender.go#L617)):*
> ```Go
> 
> 			case *roachpb.QueryIntentRequest, *roachpb.EndTxnRequest, *roachpb.GetRequest:
> 			// Accepted point requests that can be in batches with limit.
> ```

super nit: it'd be nice to keep indentation the same here and a few lines above.
___
*[pkg/sql/row/kv_batch_fetcher.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MV46b2u--EMBYTEsfu7:-MV46b2u--EMBYTEsfu8:b-fvuov4) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a5ac6896e887fa5682cb5306b65c94de1b3373/pkg/sql/row/kv_batch_fetcher.go#L266)):*
> ```Go
> 		// spans.
> 		for i := 1; i < len(spans); i++ {
> 			prevKey := spans[i-1].EndKey
> ```

nit: I'd do `s/prevKey/prevEndKey/g` and `s/curKey/curEndKey/g`.
___
*[pkg/sql/row/kv_batch_fetcher.go, line 272 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MV46IhG6DMjmwZPHfp_:-MV46IhG6DMjmwZPHfpa:b-8psjev) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a5ac6896e887fa5682cb5306b65c94de1b3373/pkg/sql/row/kv_batch_fetcher.go#L272)):*
> ```Go
> 			curKey := spans[i].EndKey
> 			if curKey == nil {
> 				curKey = spans[i-1].Key
> ```

I think this should be `spans[i].Key`, no?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @solongordon, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MV466tf7YxKxzWpZR8q:-MVd7w2f6vlUGzaUmpYi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a5ac6896e887fa5682cb5306b65c94de1b3373/pkg/kv/kvclient/kvcoord/dist_sender.go#L617)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

super nit: it'd be nice to keep indentation the same here and a few lines above.
</blockquote></details>

Done.
___
*[pkg/sql/row/kv_batch_fetcher.go, line 272 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MV46IhG6DMjmwZPHfp_:-MVd8-Ao9VXm-AkKKRYI:bdo7gwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a5ac6896e887fa5682cb5306b65c94de1b3373/pkg/sql/row/kv_batch_fetcher.go#L272)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think this should be `spans[i].Key`, no?
</blockquote></details>

Done. Thanks for noticing, this was broken.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 35 of 36 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @nvanbenschoten, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 883 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MVe-OaW12fv6Qny1Zsa:-MVe-OaX5JLZw5VmZzIF:b-jg63h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/044dab1118d79a7254ba1ef8e18bb994cdb7dc03/pkg/sql/distsql_physical_planner.go#L883)):*
> ```Go
> 		span := spans[i]
> 		if len(span.EndKey) == 0 {
> 			span = roachpb.Span{
> ```

Can you add a comment why we need to do this?
___
*[pkg/sql/span_builder_test.go, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MVe-gyZEWojsT9IWySV:-MVe-gyZEWojsT9IWySW:b-3fyp5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/044dab1118d79a7254ba1ef8e18bb994cdb7dc03/pkg/sql/span_builder_test.go#L67)):*
> ```Go
> 			prefixLen:         2,
> 			numNeededFamilies: 1,
> 			canSplit:          true,
> ```

Why is there a change here? Do we now split in more cases than before?
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MVe0kxg6O_lIxA3WlKz:-MVe0kxg6O_lIxA3WlL-:b-ncbhhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/044dab1118d79a7254ba1ef8e18bb994cdb7dc03/pkg/sql/physicalplan/fake_span_resolver.go#L82)):*
> ```Go
> 	}
> 
> 	/*
> ```

Leftover?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 883 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MVe-OaW12fv6Qny1Zsa:-MWAy-Nb7bgzi85R5TBj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044dab1118d79a7254ba1ef8e18bb994cdb7dc03/pkg/sql/distsql_physical_planner.go#L883)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Can you add a comment why we need to do this?
</blockquote></details>

Done.
___
*[pkg/sql/span_builder_test.go, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MVe-gyZEWojsT9IWySV:-MWAyH2D3jDjwKOFAtDb:b-fy6p3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/044dab1118d79a7254ba1ef8e18bb994cdb7dc03/pkg/sql/span_builder_test.go#L67)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why is there a change here? Do we now split in more cases than before?
</blockquote></details>

Yes, the idea here is that we now "split" in all cases where we can produce a single-key lookup. This example has just a single column family to look up, so we "split" the span into a single GetRequest for that column family.
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MVe0kxg6O_lIxA3WlKz:-MWAyZFBDM0opFJ57j6h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044dab1118d79a7254ba1ef8e18bb994cdb7dc03/pkg/sql/physicalplan/fake_span_resolver.go#L82)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Leftover?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 9 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @nvanbenschoten, @solongordon, and @yuzefovich)

___
*[pkg/sql/colfetcher/colbatch_scan.go, line 278 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MWl4vPk79tVCBrxV7k7:-MWl4vPk79tVCBrxV7k8:biqr71d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33e395f4c1936e57f4ab2dceb75363a7d759514/pkg/sql/colfetcher/colbatch_scan.go#L278)):*
> ```Go
> 	spans := s.spans[:0]
> 	specSpans := spec.Spans
> 	// We need to check that, if this scan is for a single row, and this scan
> ```

This seems like a leftover comment?
___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1771 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MWl57eW7ZWxDn_UEee7:-MWl57eW7ZWxDn_UEee8:b-24w3g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33e395f4c1936e57f4ab2dceb75363a7d759514/pkg/sql/opt/exec/execbuilder/testdata/select#L1771)):*
> ```
> query T
> SELECT message FROM [SHOW TRACE FOR SESSION]
> WHERE message LIKE 'querying next range% at /Table/74/1%' OR
> ```

nit: is the first `%` needed?


<!-- Sent from Reviewable.io -->

jordanlewis committed:
batcheval: respect bytes/key limits in GetRequests

Previously, bytes/key limits were ignored in GetRequest. Now, they are
respected.

Release note: None

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @solongordon, and @yuzefovich)

___
*[pkg/sql/colfetcher/colbatch_scan.go, line 278 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MWl4vPk79tVCBrxV7k7:-MWlefWg-C06Wo261bem:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33e395f4c1936e57f4ab2dceb75363a7d759514/pkg/sql/colfetcher/colbatch_scan.go#L278)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This seems like a leftover comment?
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1771 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MWl57eW7ZWxDn_UEee7:-MWlegGTDeTTyTRYvfOW:bo6uyvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d33e395f4c1936e57f4ab2dceb75363a7d759514/pkg/sql/opt/exec/execbuilder/testdata/select#L1771)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: is the first `%` needed?
</blockquote></details>

No, fixed


<!-- Sent from Reviewable.io -->

jordanlewis commented:
@tbg or @nvanbenschoten, PTAL at the first commit, which teaches `cmd_get` to leave early if it runs out of bytes or key quota. It's entirely possible that this is undesired behavior, I have no idea :) The reason that I added it is that, now that SQL can emit mixed scan and get requests, it will receive out of order results unless something like this change is added.

Consider a 4 column family table and a query that skips an inner column family.

```
CREATE TABLE a (a INT PRIMARY KEY, b INT, c INT, d INT, FAMILY(a), FAMILY(b), FAMILY(c), FAMILY(d)
SELECT a, b, d FROM a WHERE a = 10
```

This PR handles this request by creating a ScanRequest for `a,b`, and a GetRequest for `d`. If the ScanRequest runs out of quota during `a,b`, and returns a `ResumeSpan` for just `b`, but the `GetRequest` ignores the key/bytes limit and returns a result unconditionally, then SQL ends up seeing the keys in order `a`, `d`, `b`, which is invalid and causes various things to break.

I suppose we could fix this in another way, like teaching the distsender or fetcher to be smarter somehow, but that seems inelegant - to me it feels natural that `GetRequest` would obey the bytes/keys limits the same way that `ScanRequest` does.

nvb reviewed (commented):
This is fantastic! Thanks for pushing this change along and apologies for the delayed review.

> Reviewed 4 of 36 files at r2, 2 of 11 files at r3, 38 of 38 files at r4, 36 of 36 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MXJgbRJD6ZW40qUSyKI:b5cnh46)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @solongordon, @tbg, and @yuzefovich)

___
*[pkg/kv/kvserver/replica_evaluate_test.go, line 130 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJdkSsC0pySggPAkxd:-MXJdkSsC0pySggPAkxe:b-wq54ol) ([raw file](https://github.com/cockroachdb/cockroach/blob/add9d449467fc9e9ed601aa23d2cd112a0d5bc46/pkg/kv/kvserver/replica_evaluate_test.go#L130)):*
> ```Go
> 			},
> 		}, {
> 			// A batch limited to return only one key. Throw in a Get which will come
> ```

Nice! This looks like a good change. Can we add some cases for when the Get request comes first and exceeds the byte/key limit?
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJe1J92EYIDV_rHkox:-MXJe1J92EYIDV_rHkoy:b-47v0b8) ([raw file](https://github.com/cockroachdb/cockroach/blob/add9d449467fc9e9ed601aa23d2cd112a0d5bc46/pkg/kv/kvserver/batcheval/cmd_get.go#L38)):*
> ```Go
> 	var intent *roachpb.Intent
> 	var err error
> 	if h.MaxSpanRequestKeys < 0 || h.TargetBytes < 0 {
> ```

Let's leave a comment here about how the batch must have already exhausted its key/byte limit and how this mirrors the behavior of MVCCScan with a negative `MaxKeys` or `TargetBytes`.
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJe94e9qKzK_A9jgr0:-MXJe94e9qKzK_A9jgr1:b-98wk0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/add9d449467fc9e9ed601aa23d2cd112a0d5bc46/pkg/kv/kvserver/batcheval/cmd_get.go#L38)):*
> ```Go
> 	var intent *roachpb.Intent
> 	var err error
> 	if h.MaxSpanRequestKeys < 0 || h.TargetBytes < 0 {
> ```

I think it would be cleaner to structure this as an early return, since nothing that follows is needed in this case.
___
*[pkg/kv/kvserver/batcheval/cmd_get_test.go, line 28 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJeVmz4bZ35woosIQv:-MXJeVmz4bZ35woosIQw:b-yjolqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/add9d449467fc9e9ed601aa23d2cd112a0d5bc46/pkg/kv/kvserver/batcheval/cmd_get_test.go#L28)):*
> ```Go
> 	key := roachpb.Key([]byte{'a'})
> 	_, err := Get(ctx, nil, CommandArgs{
> 		Header: roachpb.Header{TargetBytes: -1},
> ```

Can we split this into three cases, one with a negative `TargetBytes`, one with a negative `MaxSpanRequestKeys`, and one with neither set?
___
*[pkg/sql/row/kv_batch_fetcher.go, line 361 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJg0fD2wuT8B3EU-oj:-MXJg0fD2wuT8B3EU-ok:b-iikcq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce5f97f74c33f7757a893200ffa83f52db5203a4/pkg/sql/row/kv_batch_fetcher.go#L361)):*
> ```Go
> 		}
> 	} else {
> 		scans := make([]struct {
> ```

It's unfortunate that this allocation is now wasted in some cases and also that it can't be used for the `Get` case. I wonder what we can do for this. Would it be crazy to loop through the spans ahead of time (outside of this `reverse` block), count the point operations, allocate an appropriately sized `gets` slice, and then do some math to accurately compute the size of these allocations?
___
*[pkg/sql/row/kv_batch_fetcher.go, line 517 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJgLCb8cpB7tP4CJew:-MXJgLCb8cpB7tP4CJex:bw0wz09) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce5f97f74c33f7757a893200ffa83f52db5203a4/pkg/sql/row/kv_batch_fetcher.go#L517)):*
> ```Go
> 		case *roachpb.GetResponse:
> 			// TODO(jordan): do we have to worry about the IntentValue field on this
> 			// response?
> ```

No, we shouldn't need to. The value is only included for lower consistency-level reads.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Thanks Nathan, it's RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MXZl_UD6-VeNDDPIi8a:b-4wec8w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @solongordon, @tbg, and @yuzefovich)

___
*[pkg/sql/row/kv_batch_fetcher.go, line 361 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJg0fD2wuT8B3EU-oj:-MXZd1ix2XpXi1nY2-zc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce5f97f74c33f7757a893200ffa83f52db5203a4/pkg/sql/row/kv_batch_fetcher.go#L361)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's unfortunate that this allocation is now wasted in some cases and also that it can't be used for the `Get` case. I wonder what we can do for this. Would it be crazy to loop through the spans ahead of time (outside of this `reverse` block), count the point operations, allocate an appropriately sized `gets` slice, and then do some math to accurately compute the size of these allocations?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 36 of 39 files at r6, 30 of 36 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @solongordon, @tbg, and @yuzefovich)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod a few small comments.

> Reviewed 3 of 39 files at r6, 7 of 36 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MXhykD5BQZsABb5r6lE:b-78aab)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @solongordon, and @tbg)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJe94e9qKzK_A9jgr0:-MXhxmVe2QCxqUdijEmR:b3usmh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/add9d449467fc9e9ed601aa23d2cd112a0d5bc46/pkg/kv/kvserver/batcheval/cmd_get.go#L38)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think it would be cleaner to structure this as an early return, since nothing that follows is needed in this case.
</blockquote></details>

nit: can we pull this up above the variable declarations, since none of the variables are used in this branch?
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 56 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXhxz3C4JIsyoATkdMB:-MXhxz3C4JIsyoATkdMC:bkc3whi) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e26475dc8349354a3247eb35ce04d1691ba818/pkg/kv/kvserver/batcheval/cmd_get.go#L56)):*
> ```Go
> 		LocalUncertaintyLimit: cArgs.LocalUncertaintyLimit,
> 	})
> 	if val != nil {
> ```

nit: we can pull this below the error handling branch.
___
*[pkg/kv/kvserver/batcheval/cmd_get_test.go, line 93 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXhyBCs28JX4Keq4Jki:-MXhyBCs28JX4Keq4Jkj:b-q4sgbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e26475dc8349354a3247eb35ce04d1691ba818/pkg/kv/kvserver/batcheval/cmd_get_test.go#L93)):*
> ```Go
> 	assert.Nil(t, resp.ResumeSpan)
> 	assert.NotNil(t, resp.Value)
> 	assert.Equal(t, resp.Value.RawBytes, value.RawBytes)
> ```

Can we also check the NumKeys and NumBytes here?
___
*[pkg/sql/row/kv_batch_fetcher.go, line 345 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXhyX8n3LwEKKgb1Olm:-MXhyX8n3LwEKKgb1Oln:b-3uk9jl) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e26475dc8349354a3247eb35ce04d1691ba818/pkg/sql/row/kv_batch_fetcher.go#L345)):*
> ```Go
> 		}
> 	}
> 	gets := make([]struct {
> ```

A zero-length slice doesn't result in a heap allocation, right?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Thanks @nvanbenschoten! Does anyone else from @cockroachdb/sql-queries want to take a look? If not, I'll merge later this week.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MXjGoeu-rPwRf8QsoZx:b-e0ott3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten, @solongordon, @tbg, and @yuzefovich)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXJe94e9qKzK_A9jgr0:-MXjDLKt5lkWrJgTKCRb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/add9d449467fc9e9ed601aa23d2cd112a0d5bc46/pkg/kv/kvserver/batcheval/cmd_get.go#L38)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: can we pull this up above the variable declarations, since none of the variables are used in this branch?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 56 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXhxz3C4JIsyoATkdMB:-MXjDLviB8Radyvva1lx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e26475dc8349354a3247eb35ce04d1691ba818/pkg/kv/kvserver/batcheval/cmd_get.go#L56)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: we can pull this below the error handling branch.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/batcheval/cmd_get_test.go, line 93 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXhyBCs28JX4Keq4Jki:-MXjE1-X7qdBQ23AV0E1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e26475dc8349354a3247eb35ce04d1691ba818/pkg/kv/kvserver/batcheval/cmd_get_test.go#L93)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we also check the NumKeys and NumBytes here?
</blockquote></details>

Done.
___
*[pkg/sql/row/kv_batch_fetcher.go, line 345 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXhyX8n3LwEKKgb1Olm:-MXjFM4KA22xmrBBD5dE:bz22sn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e26475dc8349354a3247eb35ce04d1691ba818/pkg/sql/row/kv_batch_fetcher.go#L345)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

A zero-length slice doesn't result in a heap allocation, right?
</blockquote></details>

Correct. https://github.com/golang/go/blob/master/src/runtime/malloc.go#L907-L909


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> thanks Jordan!

> Reviewed 38 of 38 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MXjK0jpBzr8ulhi3IQ-:bclj1vd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @solongordon and @tbg)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region, line 1477 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXmP7Oz0FHzpcnnJNv_:-MXmP7P-0Ypy2PCa0JKy:ba9l363) ([raw file](https://github.com/cockroachdb/cockroach/blob/1634b6ae3fb18968327b92b07022dca76618463e/pkg/ccl/logictestccl/testdata/logic_test/multi_region#L1477)):*
> ```
>               spans: [/'ap-southeast-2'/'94e4b847-8f2f-4ac5-83f1-641d6e3df727' - /'ap-southeast-2'/'94e4b847-8f2f-4ac5-83f1-641d6e3df727'] [/'us-east-1'/'94e4b847-8f2f-4ac5-83f1-641d6e3df727' - /'us-east-1'/'94e4b847-8f2f-4ac5-83f1-641d6e3df727']
> ·
> Diagram: https://cockroachdb.github.io/distsqlplan/decode.html#eJykUttO20AQfe9XjOaFizZKfIEES0hGbUqD0oQmVK3UtZDxjmHbZNfdXauuUJ76bf2vyjYUghoE9MHWzuXMnHM012i_LzDC4efT8dFoAttvRvOz-YfxDsyH4-HrM9BGkDmXgkFh9FLfvGRGcDRvH-c_pLvSpTtvGm6ruyCkzXSp3Hpj2wVvZ9P37WwLJ9PRpJ1u4eN8NDmG7dtdO_Dp3XA2_MsCDmHrIKTwYhD2O4Pczzthmu11BkHudfZDT-xTIPK-399ChkoLmqRLshh9wRAThoXRGVmrTZ26bhpGosKox1CqonR1OmGYaUMYXaOTbkEY4Vl6saAZpYJMt4cMBblULpqxrYK4MHKZmp_IcF6kykbQ5Rhz7HLkvDoIOa-o_l0MjjmvBnn35PcvzqvcE5xXnlCHddDf4tjtdSBVAjzQ7ooMMpyWLoLYY7HP4gCTFUNdujuq1qWXhJG3Yi-T4z1ZTk289zJJT5bhb5Rxx75UDUkSa8yT1T-ETnRHF11_XeJYLqUDbyOH3nOsPNFS3TgZrK9pz_mek2Otv5UFfNVSgVYRxDVgRkqQaZyBOGAQ-_W3G-9tpBc8h96MbKGVpQdWbRKeMCRxSa3fVpcmo1Ojs2ZNG04bXJMQZF1b9dtgpJpSc4r3wd7_gP1HwcEauPcQHDwKDh-Ak9WrPwEAAP__jsyklQ==
> ```

The spans are deceiving, they look like they are unbounded to the right.. It would be great if we could print just the span in this case, without `-`.


<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql: show single-key spans without a "range dash"

Previously, a single-key span (a roachpb.Span with only a StartKey and
without an EndKey) was displayed like a normal span, without anything
after a dash. For example, a single-key span starting at
`/Table/11/1/foo` would display as `/Table/11/1/foo-`.

Now, a single-key span is shown without a dash.

Release note (sql change): single-key spans in explain and
explain(distsql) are now shown without a misleading dash after them.

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten, @RaduBerinde, @solongordon, and @tbg)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region, line 1477 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXmP7Oz0FHzpcnnJNv_:-MXnO6hV3ObGNtQnQDeY:b68ik1t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1634b6ae3fb18968327b92b07022dca76618463e/pkg/ccl/logictestccl/testdata/logic_test/multi_region#L1477)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The spans are deceiving, they look like they are unbounded to the right.. It would be great if we could print just the span in this case, without `-`.
</blockquote></details>

Fixed. I added this change as a separate commit.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
@RaduBerinde PTAL

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MYDba8K-oXAX2Y7xpSf:bcmw7ss)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @jordanlewis, @nvanbenschoten, @solongordon, and @tbg)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region, line 1477 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MXmP7Oz0FHzpcnnJNv_:-MYD_zQo3x8E4csWUir4:bwl71vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1634b6ae3fb18968327b92b07022dca76618463e/pkg/ccl/logictestccl/testdata/logic_test/multi_region#L1477)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Fixed. I added this change as a separate commit.
</blockquote></details>

Thanks, looks great!
___
*[pkg/sql/rowenc/index_encoding.go, line 387 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MYDbcb67ZOKfqlWMw9k:-MYDbcb67ZOKfqlWMw9l:barv5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/475d5510050f4bc00740deb39074feabf2d4fa49/pkg/sql/rowenc/index_encoding.go#L387)):*
> ```Go
> // input span is for a table that has just a single column family, so that the
> // caller can have a precise key to send via a GetRequest if desired.
> //
> ```

[nit] worth mentioning that the returned Spans that represent a single key would not have `EndKey=nil`.
___
*[pkg/sql/rowenc/index_encoding.go, line 389 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MYDbJ7T-e4WAsivzO11:-MYDbJ7T-e4WAsivzO12:bnimo6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/475d5510050f4bc00740deb39074feabf2d4fa49/pkg/sql/rowenc/index_encoding.go#L389)):*
> ```Go
> //
> // The function accepts a slice of spans to append to.
> func SplitSpanIntoFamilySpans(
> ```

[nit] It would be a bit more precise if we passed just the Key here instead of a Span.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @jordanlewis, @nvanbenschoten, @RaduBerinde, @solongordon, and @tbg)

___
*[pkg/sql/rowenc/index_encoding.go, line 387 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MYDbcb67ZOKfqlWMw9k:-MYFaBK21ZVqC_uVk86m:bg5kz8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/475d5510050f4bc00740deb39074feabf2d4fa49/pkg/sql/rowenc/index_encoding.go#L387)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] worth mentioning that the returned Spans that represent a single key would not have `EndKey=nil`.
</blockquote></details>

Actually, returned spans that represent a single key will have nil EndKeys! I've clarified in the comment.
___
*[pkg/sql/rowenc/index_encoding.go, line 389 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MYDbJ7T-e4WAsivzO11:-MYFaerN0dS4YHi4gyD4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/475d5510050f4bc00740deb39074feabf2d4fa49/pkg/sql/rowenc/index_encoding.go#L389)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] It would be a bit more precise if we passed just the Key here instead of a Span.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTRs!

bors r+

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-:-MYLE6HS9Fqp8PdPuABA:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained

___
*[pkg/sql/rowenc/index_encoding.go, line 387 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/61583#-MYDbcb67ZOKfqlWMw9k:-MYLDup60obNXar67D9b:bggaqg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/475d5510050f4bc00740deb39074feabf2d4fa49/pkg/sql/rowenc/index_encoding.go#L387)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Actually, returned spans that represent a single key will have nil EndKeys! I've clarified in the comment.
</blockquote></details>

Yeah, I don't know what that `not` was doing in there, I gotta stop editing my comments haha


<!-- Sent from Reviewable.io -->
